build/commands/lib/affectedTests.js

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+const { promisify } = require('util')
+const { readFile, writeFile } = require('fs/promises')
+const exec = promisify(require('child_process').execFile)
+const path = require('path')
+const config = require('./config')
+const { unlink } = require('fs-extra')
+const { randomUUID } = require('crypto')
+const {
+  getApplicableFilters,
+  getTestsToRun,
+  targetToExecutable,
+} = require('./testUtils')
+const { tmpdir } = require('os')
+
+const getTestTargets = (outDir, filters = ['//*']) => {
+  const { env, shell } = config.defaultOptions
+  return exec(
+    'gn',
+    ['ls', outDir, '--type=executable', '--testonly=true', ...filters],
+    { env, shell },
+  ).then((x) => x.stdout.trim().split('\n'))
+}
+
+const asGnTarget = (file) => {
+  const gnAbsolutePath = file.startsWith('//') ? file : '//brave/' + file
+
+  return gnAbsolutePath.replace('brave/chromium_src/', '')
+}
+
+async function getModifiedFiles(target = 'HEAD~', base = null) {
+  const args = ['diff', '--name-only', target, base].filter((x) => x)
+  const maxBuffer = 1024 * 1024 * 5
+  return exec('git', args, { maxBuffer }).then((x) =>
+    x.stdout
+      .trim()
+      .split('\n')
+      .filter((x) => x),
+  )
+}
+
+async function getReferenceCommit() {
+  const currentBranch = await exec('git', [
+    'rev-parse',
+    '--abbrev-ref',
+    'HEAD',
+  ]).then((x) => x.stdout)
+
+  if (currentBranch !== 'master') {
+    return 'origin/master'
+  }
+
+  // compare with previous commit
+  return 'HEAD~'
+}
+
+async function analyzeAffectedTests(
+  outDir,
+  { filters = ['//*'], files = [], base, quiet } = {},
+) {
+  if (!quiet) {
+    console.warn('using analyzeAffectedTests is experimental')
+  }
+
+  const targetCommit =
+    !base || base === true ? await getReferenceCommit() : base
+
+  const testTargets = await getTestTargets(outDir, filters)
+  const modifiedFiles = [
+    ...(await getModifiedFiles(targetCommit)),
+    ...files,
+  ].map(asGnTarget)
+
+  const toAnalyze = {
+    files: modifiedFiles,
+    test_targets: testTargets,
+  }
+
+  // We currently don't reason about patches
+  // Let's just assume that everything needs to re-run if they are changed
+  // TODO: analyze impact of patch files
+  if (modifiedFiles.find((x) => x.startsWith('//brave/patches'))) {
+    if (!quiet) {
+      console.warn(
+        'a patch file has been modified! assuming all tests need to run',
```

Comment:
is there a follow-up issue for this? Please add `TODO(issue_link) description`

Reply:
added: https://github.com/brave/devops/issues/14212